example-2/evm/src/contracts/lendingSpoke/Spoke.sol

```diff
@@ -4,33 +4,108 @@ pragma solidity ^0.8.0;
 import "../../interfaces/IWormhole.sol";
 
 import "./SpokeSetters.sol";
-import "./SpokeStructs.sol";
-import "./SpokeMessages.sol";
+import "../lendingHub/HubStructs.sol";
+import "../lendingHub/HubMessages.sol";
 import "./SpokeGetters.sol";
 
-contract Spoke is SpokeSetters, SpokeGetters, SpokeStructs, SpokeMessages {
-    constructor(uint16 chainId_, address wormhole_, address tokenBridge_) {
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+contract Spoke is HubStructs, HubMessages, SpokeSetters, SpokeGetters {
+    constructor(uint16 chainId_, address wormhole_, address tokenBridge_, uint16 hubChainId_, address hubContractAddress) {
         setOwner(_msgSender());
         setChainId(chainId_);
         setWormhole(wormhole_);
         setTokenBridge(tokenBridge_);
+        setHubChainId(hubChainId_);
+        setHubContractAddress(hubContractAddress);
+    }
+
+    function completeRegisterAsset(bytes calldata encodedMessage) public {
+        bytes memory vmPayload = tokenBridge().completeTransferWithPayload(encodedMessage);
+        RegisterAssetPayload memory params = decodeRegisterAssetPayload(vmPayload);
+        allowAsset(params.assetAddress);
+        AssetInfo memory info = AssetInfo({
+            collateralizationRatio: params.collateralizationRatio,
+            reserveFactor: params.reserveFactor,
+            pythId: params.pythId,
+            decimals: params.decimals,
+            exists: true
+        });
+        registerAssetInfo(params.assetAddress, info);        
+    }
+
+    function depositCollateral(address assetAddress, uint256 assetAmount) public {
+        PayloadHeader memory payloadHeader = PayloadHeader({
+            payloadID: 1,
+            sender: address(this)
+        });
+
+        DepositPayload memory depositPayload = DepositPayload({
+            header: payloadHeader,
+            assetAddress: assetAddress,
+            assetAmount: assetAmount
+        });
+
+        // create WH message
+        bytes memory serialized = encodeDepositPayload(depositPayload);
+
+        sendTokenBridgeMessage(assetAddress, assetAmount, serialized);
```

Comment:
you want to do an actual token transfer here too, where the user sends tokens to the protocol pool